shared/feature_flags/feature_checker.py

```diff
@@ -0,0 +1,123 @@
+from shared.settings import settings
+from shared.database.user import User
+from shared.database.account.plan_template import PlanTemplate
+from shared.database.account.plan import Plan
+from sqlalchemy.orm.session import Session
+from shared.database.project import Project
+from shared.shared_logger import get_shared_logger
+
+logger = get_shared_logger()
+
+
+class FeatureChecker:
+    """
+        This class will check if any of the feature flags are available for a given
+        user on a give Diffgram Installation.
+
+        Eventually this class will be using an external feature flag system SDK.
+    """
+
+    user: User
+    session: Session
+    project: Project
+    install_fingerprint: str
+    FEATURE_FLAGS: list  # Temp list while feature flag system is implemented.
+
+    def __init__(self, session: Session, user: User, project: Project):
+        self.user = user
+        self.project = project
+        self.session = session
+        self.install_finger_print = settings.DIFFGRAM_INSTALL_FINGERPRINT
+        # This Dict will eventually be replaced by calls to our feature flag system.
+        self.FEATURE_FLAGS = [
+            'FREE_TIER__MAX_VIDEOS_PER_DATASET',
+            'FREE_TIER__MAX_USERS_PER_PROJECT',
+            'FREE_TIER__MAX_IMAGES_PER_DATASET',
+            'FREE_TIER__MAX_TEXT_FILES_PER_DATASET',
+            'FREE_TIER__MAX_SENSOR_FUSION_FILES_PER_DATASET',
+            'FREE_TIER__MAX_FRAMES_PER_VIDEO',
+            'FREE_TIER__MAX_INSTANCES_PER_EXPORT',
+            'FREE_TIER__MAX_PROJECTS',
+        ]
+
+    def get_or_create_free_plan(self):
+        plan_template = PlanTemplate.get_by_public_name(
+            session = self.session,
+            public_name = 'Free'
+        )
+        if not plan_template:
+            plan_template = PlanTemplate.create_free_plan(session = self.session)
+
+        plan = Plan.new(
+            session = self.session,
+            member = self.user.member,
+            plan_template = plan_template,
+            premium_plan_user_count = -1,
+            is_annual_pricing = False,
+            calculated_charge = -1,
+            per_user_final = -1,
+            marketing_promo_code = '',
+            marketing_promo_rate_found = -1,
+            marketing_plan_rate = -1,
+            marketing_savings = -1,
+            marketing_total = -1,
+            roi_monthly_engineering_cost = -1,
+            roi_monthly_supervisor_cost = -1,
+            roi_project_value = -1,
+            roi_eng_automation_benefit = -1,
+            roi_annotation_productivity_benefit = -1,
+            roi_project_quality_improvement = -1,
+            roi_total_benefit = -1,
+            roi_multiple = -1
+
+        )
+        return plan
+
+    def get_flag(self, flag_name):
+        if not settings.ALLOW_PLANS:
```

Comment:
What I'm picturing for open source part

1. We have a really clear "is open source" flag
2. Then we have a clear "open source max user" variable (eg = 20)
3. Then all the various feature flag checkers

    1. eg if open source true and not user count, then returns without doing anything (allow all). In future if we need to add stuff here to prevent abuse we can.

    2. if it's user related, then it checks "open source max user" the user count, but that's it

If we can build this into the feature_checker level always then it can help make it clear we aren't limiting people in open source context.

Sorry in advance if this is already in the code and I missed it

So maybe `IS_OPEN_SOURCE`. defaults to True, and then for diffgram.com saas we can set to false?

Potentially `ENABLE_PLANS` maybe still is a good idea to have as a feature flag just in case there is say a bug in a plan and we want to disable it for customer service reasons etc.

Reply:
Those are good ideas Anthony! Thanks for sharing

I was thinking in a similar way but not having the `IS_OPEN_SOURCE` I was thinking that all open source instalation will have the `ALLOW_PLANS` and `ALLOW_BILLING` as False with the exception of difffgram.com. That way all user install will never do any plan checks, but maybe the `IS_OPEN_SOURCE` is a bit more explicit. We can discuss more details on the call!